[Form] Add `FormFlow` for multistep forms management

yceruto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

Alternative to
* https://github.com/symfony/symfony/pull/59548

Inspired on @silasjoisten's work and @craue's [CraueFormFlowBundle](https://github.com/craue/CraueFormFlowBundle), thank you!

# FormFlow

This PR introduces `FormFlow`, a kind of super component built on top of the existing `Form` architecture. It handles the definition, creation, and handling of multistep forms, including data management, submit buttons, and validations across steps.

![formflow](https://github.com/user-attachments/assets/8a60a447-e43e-4ea5-8d77-2351522ac95a)

Demo app: https://github.com/yceruto/formflow-demo
Slides: https://speakerdeck.com/yceruto/formflow-build-stunning-multistep-forms

## `AbstractFlowType`

Just like `AbstractType` defines a single form based on the `FormType`, `AbstractFlowType` can be used to define a multistep form based on `FormFlowType`.

```php
class UserSignUpType extends AbstractFlowType
{
    public function buildFormFlow(FormFlowBuilderInterface $builder, array $options): void
    {
        $builder->addStep('personal', UserSignUpPersonalType::class);
        $builder->addStep('professional', UserSignUpProfessionalType::class);
        $builder->addStep('account', UserSignUpAccountType::class);

        $builder->add('navigator', NavigatorFlowType::class);
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            'data_class' => UserSignUp::class,
            'step_property_path' => 'currentStep', // declared in UserSignUp::$currentStep
        ]);
    }
}
```

The step name comes from the first param of `addStep()`, which matches the form name, like this:
 * The `personal` form of type `UserSignUpPersonalType` will be the step `personal`,
 * The `professional` form of type `UserSignUpProfessionalType` will be the step `professional`,
 * and so on.

When the form is created, the `currentStep` value determines which step form to build, only the matching one, from the steps defined above, will be built.

## Controller

Use the existent `createForm()` in your controller to create a `FormFlow` instance.

```php
class UserSignUpController extends AbstractController
{
    #[Route('/signup')]
    public function __invoke(Request $request): Response
    {
        $flow = $this->createForm(UserSignUpType::class, new UserSignUp())
            ->handleRequest($request);

        if ($flow->isSubmitted() && $flow->isValid() && $flow->isFinished()) {
            // do something with $form->getData()

            return $this->redirectToRoute('app_signup_success');
        }

        return $this->render('signup/flow.html.twig', [
            'form' => $flow->getStepForm(),
        ]);
    }
}
```

This follows the classic form creation and handling pattern, with 2 key differences:
 * The check `$flow->isFinished()` to know if form flow was marked as finished (when the finish flow button was clicked),
 * The `$flow->getStepForm()` call, which creates a new step form, when necessary, based on the current state.

Don't be misled by the `$flow` variable name, it's just a `Form` descendant with `FormFlow` capabilities.

> [!IMPORTANT]
>The form data will be stored across steps, meaning the initial data set during the FormFlow creation won't match the one returned by `$form->getData()` at the end. Therefore, _always_ use `$form->getData()` when the flow finishes.

## `ButtonFlowType`

A FlowButton is a regular submit button with a handler (a callable). It mainly handles step transitions but can also run custom logic tied to your form data.

There are 4 built-in Flow button types:
 * `ResetFlowType`: sends the FormFlow back to the initial state (will depend on the initial data),
 * `NextFlowType`: moves to the next step,
 * `PreviousFlowType`: goes to a previous step,
 * `FinishFlowType`: same as `reset` but also marks the FormFlow as finished.

You can combine these options of these buttons for different purposes, for example:
 * A `skip` button using the `NextFlowType` and `clear_submission = true` moves the FormFlow forward while clearing the current step,
 * A `back_to` button using the `PreviousFlowType` and a view value (step name) returns to a specific previous step,

Built-in flow buttons will have a default handler, but you can define a custom handler for specific needs. The `handler` option uses the following signature:

```php
function (UserSignUp $data, ButtonFlowInterface $button, FormFlowInterface $flow) {
    // $data is the current data bound to the form the button belongs to,
    // $button is the flow button clicked,
    // $flow is the FormFlow that the button belongs to, $flow->moveNext(), $flow->movePrevious(), ...
}
```

> [!IMPORTANT]
>By default, the callable handler is executed when the form is submitted, passes validation, and just before the next step form is created during `$flow->getStepForm()`. To control it manually, check if `$flow->getClickedButton()` is set and call `$flow->getClickedButton()->handle()` after `$flow->handleRequest($request)` where needed.

`ButtonFlowType` also comes with other 2 options:
 * `clear_submission`: If true, it clears the submitted data. This is especially handy for `skip` and `previous` buttons, or anytime you want to empty the current step form submission.
 * `include_if`: `null` if you want to include the button in all steps (default), an array of steps, or a callable that’s triggered during form creation to decide whether the flow button should be included in the current step form. This callable will receive the `FormFlowCursor` instance as argument.

## Other Building Blocks

<details>
<summary><h4>FormFlowCursor</h4></summary>

This immutable value object holds all defined steps and the current one. You can access it via `$flow->getCursor()` or as a `FormView` variable in Twig to build a nice step progress UI.

</details>

<details>
<summary><h4>NavigatorFlowType</h4></summary>

The built-in `NavigatorFlowType` provides 3 default flow buttons: `previous`, `next`, and `finish`. You can customize or add more if needed. Here’s an example of adding a “skip” button to the `professional` step we defined earlier:

```php
class UserSignUpNavigatorType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder->add('skip', NextFlowType::class, [
            'clear_submission' => true,
            'include_if' => ['professional'], // the step names where the button will appear
        ]);
    }

    public function getParent(): string
    {
        return NavigatorFlowType::class;
    }
}
```
Then use `UserSignUpNavigatorType` instead.
</details>

<details>
<summary><h4>Data Storage</h4></summary>

FormFlow handles state across steps, so the final data includes everything collected throughout the flow. By default, it uses `SessionDataStorage` (unless you’ve configured a custom one). For testing, `InMemoryDataStorage` is also available.

You can also create custom data storage by implementing `DataStorageInterface` and passing it through the `data_storage` option in `FormFlowType`.

</details>

<details>
<summary><h4>Step Accessor</h4></summary>

The `step_accessor` option lets you control how the current step is read from or written to your data. By default, `PropertyPathStepAccessor` handles this using the form’s bound data and `PropertyAccess` component. If the step name is managed externally (e.g., by a workflow), you can create a custom `StepAccessorInterface` adapter and pass it through this option in `FormFlowType`.

</details>

<details>
<summary><h4>Validation</h4></summary>

FormFlow relies on the standard validation system but introduces a useful convention: it sets the current step as an active validation group. This allows step-specific validation rules without extra setup:

```php
final class FormFlowType extends AbstractFlowType  
{
    public function configureOptions(OptionsResolver $resolver): void  
    {
        // ...

        $resolver->setDefault('validation_groups', function (FormFlowInterface $flow) {  
            return ['Default', $flow->getCursor()->getCurrentStep()];  
        });
    }
}
```

Allowing you to configure the validation `groups` in your constraints, like this:

```php
class UserSignUp
{
    public function __construct(
        #[Valid(groups: ['personal'])]
        public Personal $personal  = new Personal(),

        #[Valid(groups: ['professional'])]
        public Professional $professional = new Professional(),

        #[Valid(groups: ['account'])]
        public Account $account = new Account(),

        public string $currentStep = 'personal',
    ) {
    }
}
```
</details>

<details>
<summary><h4>Type Extension</h4></summary>

FormFlowType is a regular form type in the Form system, so you can use `AbstractTypeExtension` to extend one or more of them:

```php
class UserSignUpTypeExtension extends AbstractTypeExtension
{
    /**
     * @param FormFlowBuilderInterface $builder
     */
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder->addStep('role', UserSignUpRoleType::class, priority: 1); // added to the beginning cos higher priority
        $builder->removeStep('account');
        if ($builder->hasStep('professional')) {
            $builder->getStep('professional')->setSkip(fn (UserSignUp $data) => !$data->personal->working);
        }
        $builder->addStep('onboarding', UserSignUpOnboardingType::class); // added at the end
    }

    public static function getExtendedTypes(): iterable
    {
        yield UserSignUpType::class;
    }
}
```
</details>

---

There’s a lot more to share about this feature, so feel free to ask if anything isn’t clear.

Cheers!

yceruto committed:
Add FormFlow for multistep forms management

94noni commented on src/Symfony/Component/Form/Extension/Core/Type/FormFlowNavigatorType.php:
```diff
@@ -0,0 +1,52 @@
+{
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $builder->add('back', FormFlowActionType::class, [
```
what bout "previous" ? generally it goes with "next"
or rename other "forward" ?

yceruto replied:
I don’t have a strong opinion on that. I personally prefer the shorter version, and users can change it using the `label` option anyway. Let’s see what others think.

94noni replied:
copy that :)

smnandre replied:
I have no preference between "prev/next" and "back/forward" but it does feel weird to have both, so i would agree with @94noni here.

In multiple steps form i do believe "next" is some de-facto standard, so "previous" would be the one ?

yceruto replied:
Ok, I will rename the default name/label to "Previous" instead of "Back"

yceruto replied:
Updated!

94noni commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/form.php:
```diff
@@ -123,6 +124,10 @@
             ->args([service('form.type_extension.form.request_handler')])
             ->tag('form.type_extension')
 
+        ->set('form.type_extension.form.flow.session_data_storage', FormFlowTypeSessionDataStorageExtension::class)
```
I like that this storage (default to session) is internal since day one, will help a lot to avoid ppl mistake I think

yceruto replied:
Yep, new data storages can also be added later.

94noni commented on src/Symfony/Component/Form/Extension/Core/Type/FormFlowType.php:
```diff
@@ -0,0 +1,105 @@
+use Symfony\Component\PropertyAccess\PropertyPathInterface;
+
+/**
+ * A multistep form.
```
a question about naming:

you describe this as a form multistep 
but the class is named form flow

its feels weird at first read

yceruto replied:
They mean the same to me... I’m just phrasing it differently to make the idea clearer. What is the weird part?

94noni replied:
"flow" vs "multistep" was what I meant
multistep is clear to me, I do also prefer this naming over for class name type

94noni commented on src/Symfony/Component/Form/Extension/HttpFoundation/Type/FormFlowTypeSessionDataStorageExtension.php:
```diff
@@ -0,0 +1,46 @@
+            return;
+        }
+
+        $key = \sprintf('_form_flow.%s_%s', strtolower(str_replace('\\', '_', $builder->getType()->getInnerType()::class)), $builder->getName());
```
what about prefix with `_sf` like the http session attributes ?

yceruto replied:
I think it's already the case, all Symfony session attributes are currently stored under `_sf2_attributes` key by default.

94noni commented on src/Symfony/Component/Form/Extension/Core/Type/FormFlowActionType.php:
```diff
@@ -0,0 +1,63 @@
+            ->default(null)
+            ->allowedTypes('null', 'callable');
+
+        $resolver->define('include_if')
```
nice

yceruto replied:
I've improved it a bit more for better DX, so now we have three allowed values:
 1. `null` if you want to include the button in all defined steps,
 2. an `array` with the names of the steps where the button will be included, (most used)
 3. a `callable` for more advanced inclusion logic.

94noni commented on src/Symfony/Component/Form/Extension/Core/Type/FormFlowNavigatorType.php:
```diff
@@ -0,0 +1,52 @@
+            'include_if' => fn (FormFlowCursor $cursor): bool => $cursor->canMoveNext(),
+        ]);
+
+        $builder->add('finish', FormFlowActionType::class, [
```
"last" ?

according to previous/next, last would be in the same naming logic, finish would more likely to be an action

yceruto replied:
Same as https://github.com/symfony/symfony/pull/60212#discussion_r2045260968

yceruto replied:
Ah btw, it's actually an action :) so `finish` is better than `last`, note that they are submit buttons

94noni reviewed (commented):
early raw reviews :)

connorhu commented:
Great idea! I've looked at the previous PR and could use it! One question came up: how can I jump back to a specific step without having to press the back button many times? 
May the answer is can't by default. I solved this issue by keeping track of which valid step we are at and the url contains the step number (1..n), so I can generate a url for the previous steps, but I dont allow to go to not-yet-valid next steps.
I think I can see the points where I need to customize to achieve this.

yceruto commented:
>how can I jump back to a specific step without having to press the back button many times?

Hey! take a look at the `testMoveBackToStep()` test, it covers this case. It's possible via submit operation or manually using `$flow->moveBack('step')` directly.

stof commented:
> 1. I think it would be best to already store the file(s) to a temporary location on the filesystem and just work with references to them.

this would be incompatible with apps using a load balancer with several servers, as there is no guarantee that the next request goes to the same server behind the load balancer. Such case require storing uploads in a shared storage (for instance using a S3 bucket or similar storage)

yceruto commented:
I haven’t checked the file upload yet because I knew it would be a complicated topic, but I think using a custom action button (as explained above) will give you the flexibility to do custom things on your own.

One option is to create your own `next` action button with a dedicated `handler` where file uploads can be managed. Then, you can keep track of the uploaded files by saving the references in the DTO, which is preserved between steps.

Imagine you have a `documents` step defined for files uploading. In your navigator form, you can define this new action:
```php
$builder->add('upload', FormFlowActionType::class, [
    'include_if' => fn (FormFlowCursor $cursor) => 'documents' === $cursor->getCurrentStep(),
    'handler' => function (MyDto $data, ActionButtonInterface $button, FormFlowInterface $flow) {
        // handle files uploading here ... store them somewhere ... create references ...

        // $data->uploadedFiles = ... save references if we go back ...

        $flow->moveNext();
    },
]);
```
So, it’s up to you where to store the files, how to reference them in the DTO, and how to render them again if the user goes back to this step, just by looking into `$data->uploadedFiles`.

yceruto commented:
Think of FormFlow action buttons like mini-controllers with a focused job. The main controller handles the shared logic across all steps, while each action handler takes care of custom operations.

You can even use them for inter-step operations (like in the demo preview, where I showed how to add or remove skill items from a `CollectionType`). In those cases, the step stays the same, but the bound data gets updated, and the form is rebuilt when `$flow->getStepForm()` is called again, now with new data and form updated.

This part feels a bit like magic, but it’s a helpful one given how complex multistep forms can be.

yceruto commented:
>Can we remove the draft status @yceruto ?

It's officially ready for review now :)

yceruto commented:
@seb-jean I’ve played with a `Role` entity, the `EntityType`, and the default storage (Session), and I encountered the following runtime exception when moving back to the previous step:

>Entity of type "App\Entity\Role" passed to the choice field must be managed. Maybe you forget to persist it in the entity manager?

That happens during the choice value resolution, where the `IdReader` tries to get the ID value of the entity (first checking if the given entity is currently managed) but it fails in this case (as expected) because the entity object was retrieved from the session, not from the EM (which is fresh in the next request).

In this scenario, given the context, my recommendation is to provide the `choice_value` resolver directly:
```php
->add('role', EntityType::class, [
    'class' => Role::class,
    'choice_value' => fn (?Role $role) => $role?->id, // <-- resolving the choice value manually
])
```
I don’t think `FormFlow` (or its data storage) can handle this situation any better.

Once the `FormFlow` is finished, it's over, and you've full control over the following [doctrine recommendations](https://www.doctrine-project.org/projects/doctrine-orm/en/3.3/cookbook/entities-in-session.html), either by mapping your DTO back to entities or by reloading the related entities when necessary.

yceruto commented on src/Symfony/Component/Form/Extension/Core/Type/FormFlowActionType.php:
```diff
@@ -0,0 +1,93 @@
+{
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->define('action')
```
I’m not entirely happy with the name of this option, it works like a preset configuration to set up predefined behavior, so maybe `preset` would be clearer?

Another alternative could be removing this option and creating several `FormFlowActionType` classes instead (one per action button), e.g. `FormFlowBackType`, `FormFlowNextType`, and so on. Each one with its own dedicated default options.

What do you think?

yceruto replied:
I've removed this option, so now it's split into FlowResetType, FlowPreviousType, FlowNextType, and FlowFinishType button with their specific options

yceruto reviewed (commented):
I’m still hesitating over this design topic. I’d appreciate your thoughts on this.

seb-jean commented:
@yceruto Thank you for your reply. Have you also tried to see if the `User` entity is persisted in the database once the `FormFlow` is finished?

yceruto commented on src/Symfony/Component/Form/Extension/Core/Type/FormFlowActionType.php:
```diff
@@ -0,0 +1,94 @@
+                }
+
+                return function (mixed $data, ActionButtonInterface $button, FormFlowInterface $flow): void {
+                    match (true) {
+                        $button->isBackAction() => $flow->moveBack($button->getViewData()),
+                        $button->isNextAction() => $flow->moveNext(),
+                        $button->isFinishAction() => $flow->reset(),
```
I don’t think they’re necessary.

yceruto commented on src/Symfony/Component/Form/Flow/FormFlow.php:
```diff
@@ -0,0 +1,262 @@
+                continue;
+            }
+
+            if ($child->count() > 0) {
```
It's not necessary IMO

Lorenzschaef commented:
This looks very interesting, I'm looking forward to using this component!

I recently built a wizard in my project, where the steps vary based on user choices. To take up your signup example: Imagine that in the first step, the user has to choose between signing up as a company or as an individual, and the following steps would be different based on that choice. Would something like that be possible to build with this component?

Spomky commented:
Hi @yceruto,

Thank you very much for the demo.  I found it really straightforward to understand how to use and customize the form, the steps, and the actions 👍🏽 .

I’ve tested several use cases, and everything works well except for one scenario. I might be missing something, but it seems that the form events on the steps are not being triggered. Specifically, in a flow where the user is first asked to pick a color (choice type), and then to select a gradient based on the chosen color, the gradient step uses a choice type with options depending on the previous selection. Unfortunately, the event responsible for setting these dynamic choices is never called.

I might be misunderstanding how to properly use form events with the form flow, but I wanted to check with you in case there’s a recommended way to handle this scenario.

Other than this issue, everything else is working perfectly.

Spomky commented:
Hi,

As mentioned [in the PR](https://github.com/yceruto/formflow-demo/pull/1) I created to show the issue, I eventually managed to get it working. So to me, this is definitely a very good addition, allowing lots of use cases to be covered.
Also I confirm it works fine with Doctrine entities.

Thank you very much for your work on this! I really appreciate the time and effort you put into it. Approved!

RafaelKr commented:
Would it be possible to already provide it via an experimental "addon" bundle (symfony/experimental-form) before it's released with the official Symfony Forms component? That way it could already be battle-tested with a current stable release (7.3+ or even earlier).

Waiting until 8.1 (released in May 2026) for a first experimental release seems like a pretty long time to me.

ker0x commented:
@RafaelKr there's already a bundle port of the feature that you can use: https://github.com/yceruto/formflow-bundle

The demo app use it!

yceruto commented:
I received feedback regarding the action handler logic, so I simplified it by delegating execution responsibilities directly to the `ActionButton` class. This allowed me to remove redundant methods and improve overall code clarity.

In general, you don’t need to manually call `$flow->getClickedActionButton()->handle()`, as it’s automatically executed during the `$flow->getStepForm()` process (if the clicked submit button is an ActionButton and hasn’t been handled yet).

However, you might occasionally need to invoke the handler associated with the clicked action button manually, such as when you want to perform additional logic after its execution but before the new step form:
```php
$flow->handleRequest($request);

$flow->getClickedActionButton()->handle();

// do something between the current step and the new step form creation...

$flow = $flow->getStepForm();
```

The related changes are visible [here](https://github.com/yceruto/formflow-bundle/pull/3/files).

yceruto commented:
Hey there! Just one more review & vote and we can merge this into 7.4 😊 Friendly ping to @symfony/mergers.

A quick reminder: there’s a [demo app](https://github.com/yceruto/formflow-demo) available if you’d like to test it out first. You can also check out the mirrored [bundle](https://github.com/yceruto/formflow-bundle) featuring the same code.

Thanks! 🙌

smnandre commented on src/Symfony/Component/Form/Flow/ActionButtonInterface.php:
```diff
@@ -0,0 +1,66 @@
+/**
+ * @author Yonel Ceruto <open@yceruto.dev>
+ */
+interface ActionButtonInterface extends FormInterface, ClickableInterface
```
I must admit i'm not convinced by the "action" term in here (as buttons trigger action by nature), but i see the constraint and have not much in mind to improve. 

Maybe FlowButton would be easier to document/explain ?

yceruto replied:
I like your suggestion! I wasn't fully convinced about the term "Action" either. I'll remove it from all artifacts so that we gain more consistency in the naming:
 * ActionButtonInterface -> FlowButtonInterface
 * ActionButton -> FlowButton
 * ActionButtonTypeInterface -> FlowButtonTypeInterface
 * FormFlowActionType -> FlowButtonType
 * ActionButtonBuilder -> FlowButtonBuilder

yceruto replied:
Updated!

smnandre commented on src/Symfony/Component/Form/Flow/ActionButtonInterface.php:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * Executes the callable handler.
+     */
+    public function handle(): void;
```
This feels like a lot of responsibilities for such object no? 

By reading this interface only, I would probably not understand why both getHandler() and handle() co-exist

yceruto replied:
Yes, that was part of a recent refactoring. The getHandler() method is no longer needed, so I'm removing it now…

yceruto replied:
Simplified!

yceruto commented:
Thanks @smnandre for your review!

Summary of the latest changes:
 * Renamed `back` to `previous` (wherever it makes sense)
 * Improved class naming for Flow button, type, step, builder, cursor, and navigator (now shorter and more concise)
 * Removed the `action` option from `FlowButtonType` in favor of dedicated types: `FlowResetType`, `FlowPreviousType`, `FlowNextType`, and `FlowFinishType`, each with specific options tailored to their respective actions

The description of the PR was also updated to reflect the current state of the code.

It's ready for code review again!

yceruto commented:
@virtualize I've actually planned to work on that topic today 🙂

yceruto commented:
Updated bundle and demo with latest changes.

yceruto commented:
@virtualize Yes, by default, the back/previous step action clears the submitted data from the current step form. It's like this because, from my experience, most multistep forms I've built have the current step hardly depending on the previous step's data.

See `FlowPreviousType` definition:
```php
class FlowPreviousType extends AbstractType implements FlowButtonTypeInterface
{
    // ...

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            // ...

            'clear_submission' => true, // <-- data sent will never be mapped into the current step form
        ]);
    }

    // ...
}
```
As you may guess, by setting it to `false` you'll disable that behaviour.

However, in the demo (I guess you're referring to the SignUp demo), we are assuming that going back to the previous step will mean discarding any changes made in the current step and restoring the previous step's original state. To review, you can simply click on the step number and move forward again with all the data preserved.

javiereguiluz commented:
Yonel, thank you so much for this contribution. Words can't do it justice: it's simply brilliant and wonderful ❤️ 

I love everything you're proposing, but I have some concerns about using the word "flow." The most common names for this type of forms are "multi-step form" or "stepped form" _(historically, they were also called "form wizard")_. The word "flow" in this context seems a bit vague to me.

Since you're already using the term "step" throughout much of this new functionality, why not use it everywhere? Here's how the naming would look if we used "multi-step form":

| Current Name           | Proposed Alternative
| --------------------- | --------------------------
| `AbstractFlowType`    | `AbstractMultiStepType`
| `FlowResetType`       | `ResetFormType`
| `FlowNextType`        | `NextStepType`
| `FlowPreviousType`    | `PreviousStepType`
| `FlowFinishType`      | `FinishFormType`
| `FlowNavigatorType`   | `MultiStepNavigatorType`
| `FlowButtonInterface` | `MultiStepButtonInterface`
| `FormFlowInterface`   | `MultiStepFormInterface`
| `FlowCursor`          | `MultiStepCursor`

chalasr reviewed (approved):
Huge work, love it! :shipit:

fabpot commented on src/Symfony/Component/Form/Extension/HttpFoundation/Type/FormFlowTypeSessionDataStorageExtension.php:
```diff
@@ -0,0 +1,44 @@
+
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        \assert($builder instanceof FormFlowBuilderInterface);
```
We're not using asserts in Symfony, it should throw an exception instead.

yceruto replied:
Updated, thanks!

nicolas-grekas commented on src/Symfony/Component/Form/Flow/AbstractFlowButtonType.php:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @author Yonel Ceruto <open@yceruto.dev>
+ */
+abstract class AbstractFlowButtonType extends AbstractType implements FlowButtonTypeInterface
```
should this class be in the Type namespace?
also, what about naming it AbstractButtonFlowType

yceruto replied:
I've mirrored the same structure as other form classes, for example, all existing related-type abstracts and interfaces are located directly under the `Form/` directory. Here, they are directly under `Form/Flow/`. 

Is it okay to keep it like this for consistency?

yceruto replied:
about the naming suggestion I think it makes sense yes

nicolas-grekas replied:
Makes sense!

nicolas-grekas commented on src/Symfony/Component/Form/Flow/AbstractFlowType.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * @author Yonel Ceruto <open@yceruto.dev>
+ */
+abstract class AbstractFlowType extends AbstractType implements FormFlowTypeInterface
```
to be moved inside the Type subns?

yceruto replied:
same as https://github.com/symfony/symfony/pull/60212#discussion_r2438758702 ?

nicolas-grekas commented on src/Symfony/Component/Form/Flow/Type/FlowButtonType.php:
```diff
@@ -0,0 +1,65 @@
+ *
+ * @author Yonel Ceruto <open@yceruto.dev>
+ */
+class FlowButtonType extends AbstractType implements FlowButtonTypeInterface
```
naming, would this make sense?
```suggestion
class ButtonFlowType extends AbstractType implements ButtonFlowTypeInterface
```

javiereguiluz replied:
I'm not sure about this. To me they mean different things:

* `ButtonFlow` = a flow composed of buttons
* `FlowButton` = a button that belongs to a (form) flow

chalasr replied:
The suggested new names makes FlowType a generic suffix for Flow related types, translating to "it's a Button Type from the Flow area". 
They are indeed less meaningful from a reader POV but might be more discoverable. 
That's worth challenging, I've no strong opinion though I do like names that read as a sentence (as it was before) especially for interfaces.

yceruto replied:
both options are acceptable to me, I don't have a strong opinion either way, so I'm fine with whatever we agree on

xabbuh replied:
I would stick with the current naming as that IMO is more clear about what it means.

nicolas-grekas commented on src/Symfony/Component/Form/Flow/FlowButtonTypeInterface.php:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @author Yonel Ceruto <open@yceruto.dev>
+ */
+interface FlowButtonTypeInterface extends FormTypeInterface
```
in the Type subns, as `ButtonFlowTypeInterface`?

yceruto replied:
same as https://github.com/symfony/symfony/pull/60212#discussion_r2438758702 ?

xabbuh commented on src/Symfony/Component/Form/Extension/HttpFoundation/Type/FormFlowTypeSessionDataStorageExtension.php:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        $key = \sprintf('_sf_formflow.%s_%s', strtolower(str_replace('\\', '_', $builder->getType()->getInnerType()::class)), $builder->getName());
+        $builder->setDataStorage(new SessionDataStorage($key, $this->requestStack));
```
We have two places that call `setDataStorage()` (here and in `FormFlowType`). Wouldn't it make more sense to leave the handling of the method call to the type and only set a default option value here?

yceruto replied:
Yes, that makes sense to me as well. However, I couldn’t find a way to build a unique key in that context without the $builder instance. Any idea?

xabbuh replied:
Hm, that's indeed a valid reason we cannot move this into the form type. Should we move the call from the `FormFlowType` into the `CoreExtension` and check for the return value of `getDataStorage()` above instead? I guess we need something like that already to allow for userland form type extensions to use `setDataStorage()` themselves?

yceruto replied:
if the goal is allowing for userland form type extensions to use `setDataStorage()` themselves, they can do it currently:
 * via `configureOptions()` method, then the FormFlowType will set it (and the current session extension will skip)
 * or setting it in the extension `buildForm()` method directly, it will override the current data storage as extensions are processed after the FormFlowType targeted

nicolas-grekas commented:
Thank you @yceruto.
